internal/ethapi/api.go

```diff
@@ -943,8 +949,36 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
 	// this makes sure resources are cleaned up.
 	defer cancel()
 
+	enrichedCtx := ctx
+	// create callbacks to support runtime multitenancy checks during the run
+	if authToken, ok := b.SupportsMultitenancy(ctx); ok {
+		var authorizeMessageCallFunc multitenancy.AuthorizeMessageCallFunc = func(contractAddress common.Address) (bool, bool, error) {
+			var readSecAttr *multitenancy.ContractSecurityAttribute
+			if len(data) == 0 { // public READ
+				readSecAttr = multitenancy.NewContractSecurityAttributeBuilder().FromEOA(addr).ToEOA(*msg.To()).Public().Read().Build()
+			} else {
+				currentBlock := b.CurrentBlock().Number().Int64()
+				extraDataReader, err := b.AccountExtraDataStateReaderByNumber(ctx, rpc.BlockNumber(currentBlock))
+				if err != nil {
+					return false, false, fmt.Errorf("no account extra data reader at block %v: %w", currentBlock, err)
+				}
+				managedParties, err := extraDataReader.ReadManagedParties(contractAddress)
+				isPrivate := true
+				if errors.Is(err, common.ErrNotPrivateContract) {
+					isPrivate = false
+				} else if err != nil {
+					return false, false, fmt.Errorf("%s not found in the index, error: %s", contractAddress.Hex(), err.Error())
+				}
+				readSecAttr = multitenancy.NewContractSecurityAttributeBuilder().FromEOA(addr).PrivateIf(isPrivate).PartiesOnlyIf(isPrivate, managedParties).Read().Build()
+			}
+			authorizedRead, _ := b.IsAuthorized(ctx, authToken, readSecAttr)
+			return authorizedRead, false, nil
+		}
+		enrichedCtx = context.WithValue(enrichedCtx, multitenancy.CtxKeyAuthorizeMessageCallFunc, authorizeMessageCallFunc)
```

Comment:
This looks a bit black box would there not be a cleared way to pass the `authorizeMessageCallFunc` into the EVM?

Reply:
Keeping `GetEVM` signature makes merging with upstream less conflicts. Also the logic is encapsulated inside the `GetEVM`, `EVM` has the explicit declaration of those callbacks.